truffle/src/main/ruby/core/proc_mirror.rb

```diff
@@ -0,0 +1,63 @@
+# Copyright (c) 2007-2015, Evan Phoenix and contributors
+# All rights reserved.
+#
+# Redistribution and use in source and binary forms, with or without
+# modification, are permitted provided that the following conditions are met:
+#
+# * Redistributions of source code must retain the above copyright notice, this
+#   list of conditions and the following disclaimer.
+# * Redistributions in binary form must reproduce the above copyright notice
+#   this list of conditions and the following disclaimer in the documentation
+#   and/or other materials provided with the distribution.
+# * Neither the name of Rubinius nor the names of its contributors
+#   may be used to endorse or promote products derived from this software
+#   without specific prior written permission.
+#
+# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+# AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+# IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+# DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE
+# FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+# DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+# SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+# CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+# OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+module Rubinius
+  class Mirror
+    class Proc < Mirror
+      def self.from_block(klass, env)
+        begin
+          return Rubinius.invoke_primitive :proc_from_env, env, klass
+        rescue Rubinius::Internal => exc
+          if Type.object_kind_of? env, BlockEnvironment
+            msg = "unable to create Proc from BlockEnvironment"
+            raise PrimitiveFailure, msg, exc
+          end
+        end
+
+        begin
+          env.to_proc
+        rescue Exception
+          raise ArgumentError, "Unable to convert #{env.inspect} to a Proc"
+        end
+      end
+
+      def curry(executable, args, arity)
+        args.freeze
+
+        name = executable.lambda? ? :lambda : :proc
+
+        Proc.__send__(name) do |*a, &b|
```

Comment:
Note that Proc does not have a `lambda` method, it's Kernel.
Right, this works because `send` and `__send__` are ignored, I forgot.
Sounds good then as `Kernel.__send__(name)`.